### PR TITLE
feat: Paste block in viewport / near cursor

### PR DESCRIPTION
**Pull request by bsampada7**
## The basics

- [x] I [validated my changes](https://developers.google.com/blockly/guides/contribute/core#making_and_verifying_a_change)

### Resolves

Fixes https://github.com/google/blockly/issues/6848

### Proposed Changes

When copying a block, blocky pastes the block near the original block. With this change, it will paste the new block either in the middle of the view or near where the cursor is.

### Documentation

The method getRelativeToSurfaceXY(): Coordinate is added to the IMovable interface

### CODE REVIEW

**Comment by rachel-fenichel** [comment]
Thanks for contributing! I'm working on reviewing, and I've approved the workflows to run.

File: core/interfaces/i_movable.ts [hunk 1]
```diff
@@ -4,6 +4,8 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+import {Coordinate} from '../utils';
```
Comment (rachel-fenichel) [inline core/interfaces/i_movable.ts:7]: Import from `'../utils/coordinate.js'`

File: core/shortcut_items.ts [hunk 2]
```diff
@@ -12,6 +12,7 @@ import * as common from './common.js';
 import {Gesture} from './gesture.js';
 import {ICopyData, isCopyable} from './interfaces/i_copyable.js';
 import {KeyboardShortcut, ShortcutRegistry} from './shortcut_registry.js';
+import {Coordinate} from './utils.js';
```
Comment (rachel-fenichel) [inline core/shortcut_items.ts:15]: Import from `'./utils/coordinate.js'`

File: core/interfaces/i_movable.ts (hunk 1 above)
Comment (bsampada7) [reply core/interfaces/i_movable.ts:7]: Thanks for the comment. I have changed accordingly and pushed it.

File: core/shortcut_items.ts [hunk 3]
```diff
@@ -188,8 +192,27 @@ export function registerPaste() {
       return !workspace.options.readOnly && !Gesture.inProgress();
     },
     callback() {
-      if (!copyData || !copyWorkspace) return false;
-      return !!clipboard.paste(copyData, copyWorkspace);
+      if (!copyData || !copyCoords || !copyWorkspace) return false;
+      const {
```
Comment (rachel-fenichel) [inline core/shortcut_items.ts:196]: For clarity, please update these names to `viewportLeft`, `viewportTop`, etc. That helps the next person who reads this!

File: core/shortcut_items.ts [hunk 4]
```diff
@@ -188,8 +192,27 @@ export function registerPaste() {
...
+        left: vpLeft,
+        top: vpTop,
+        width: vpWidth,
+        height: vpHeight,
+      } = copyWorkspace.getMetricsManager().getViewMetrics(true);
+      if (
```
Comment (rachel-fenichel) [inline core/shortcut_items.ts:202]: Add a comment explaining the two cases (when the default location is inside the viewport vs not).

File: core/interfaces/i_movable.ts [hunk 5]
```diff
@@ -16,4 +18,12 @@ export interface IMovable {
    * @returns True if movable.
    */
   isMovable(): boolean;
+
+  /**
+   * Return the coordinates of the top-left corner of this movable object
```
Comment (rachel-fenichel) [inline core/interfaces/i_movable.ts:23]: Since this changes the API, please change the PR title to "feat:" instead of "fix:"

**Comment by rachel-fenichel** [comment]
I've added a few small comments. Once you fix those, you should be good to go! I'll keep an eye on this to keep approving the checks as you push changes.

**Review by rachel-fenichel — APPROVED** [review]

**Comment by bsampada7** [comment]
Thanks for the merge @rachel-fenichel . Happy to have contributed on my first issue

**Comment by laurensvalk** [comment]
It looks like this was merged into the `osd` branch. Is this expected to be included in a release eventually or does this serve a different purpose? Thanks!

**Pull request merged**